README.md
# Render QR codes easily from your Rails 3 application

This gem supports rendering either SVG or PNG, JPEG, and GIF formats.

SVG, because of it's vector nature, will scale easily when intended for print. Offering QR endpoints enables others to integrate with your service in possibly interesting ways.

## Installation

Add the following to your `Gemfile`.

    gem 'rqrcode-rails3'

If you want to use the PNG, JPEG or GIF format, you will have to have **ImageMagick** installed on your system.
You will also want to add the **mini_magick** gem to your application's `Gemfile`.

    gem 'mini_magick'

## How to use

In your controller actions, you could return a QR code that links to the current page like this:

```ruby
respond_to do |format|
  format.html
  format.svg  { render :qrcode => request.url, :level => :l, :unit => 10 }
  format.png  { render :qrcode => request.url }
  format.gif  { render :qrcode => request.url }
  format.jpeg { render :qrcode => request.url }
end
```
  
#### Options:

* `:size`   – This controls how big the QR Code will be. Smallest size will be chosen by default. Set to maintain consistent size.
* `:level`  – The error correction level, can be:
  * Level `:l` 7%  of code can be restored
  * Level `:m` 15% of code can be restored
  * Level `:q` 25% of code can be restored
  * Level `:h` 30% of code can be restored (default :h) 
* `:offset` – Padding around the QR Code (e.g. 10)
* `:unit`   – How many pixels per module (e.g. 11)
* `:fill`   – Background color (e.g "ffffff" or :white)
* `:color`  – Foreground color for the code (e.g. "000000" or :black)
  
## About

This project was inspired by the first chapter in José Valim's book [Crafting Rails Applications](http://pragprog.com/titles/jvrails/crafting-rails-applications)

QR codes are encoded by [rqrcode](https://github.com/whomwah/rqrcode)

History
== 0.1.6 2012-12-04
* major enhancements
  * Added support for JPEG and GIF rendering - ([@shawnpyle][])
== 0.1.5 2012-02-10
* major enhancements
  * Added :level option
  * Added :unit size option
== 0.1.4 2012-02-09
* minor enhancements
  * #11: Check whether mime types are already defined (suppress warnings) - ([@peternash][])
  * Version of rqrcode bumped up to 0.4.2
== 0.1.3 2011-07-11
* major enhancements
  * Added :size option and default size calculator
== 0.1.2 2011-06-08
* bug fix
  * gem no longer needs to have a require statement
  * Basic color setting options implemented, however they render poorly or not at all in PNG format


test/rqrcode_svg_test.rb
require 'test_helper'

class RQRCode::SVGTest < ActiveSupport::TestCase
  test "truth" do
    assert_kind_of Module, RQRCode
  end
end


test/dummy/config/routes.rb
Dummy::Application.routes.draw do
  match "/home(.:format)", :to => "home#index", :as => :home
  match "/", :to => "home#index"
end


test/dummy/config/application.rb
require File.expand_path('../boot', __FILE__)

require "active_model/railtie"
require "active_record/railtie"
require "action_controller/railtie"
require "action_view/railtie"
require "action_mailer/railtie"

Bundler.require
require "rqrcode-rails3"

module Dummy
  class Application < Rails::Application
    # Settings in config/environments/* take precedence over those specified here.
    # Application configuration should go into files in config/initializers
    # -- all .rb files in that directory are automatically loaded.

    # Custom directories with classes and modules you want to be autoloadable.
    # config.autoload_paths += %W(#{config.root}/extras)

    # Only load the plugins named here, in the order given (default is alphabetical).
    # :all can be used as a placeholder for all plugins not explicitly named.
    # config.plugins = [ :exception_notification, :ssl_requirement, :all ]

    # Activate observers that should always be running.
    # config.active_record.observers = :cacher, :garbage_collector, :forum_observer

    # Set Time.zone default to the specified zone and make Active Record auto-convert to this zone.
    # Run "rake -D time" for a list of tasks for finding time zone names. Default is UTC.
    # config.time_zone = 'Central Time (US & Canada)'

    # The default locale is :en and all translations from config/locales/*.rb,yml are auto loaded.
    # config.i18n.load_path += Dir[Rails.root.join('my', 'locales', '*.{rb,yml}').to_s]
    # config.i18n.default_locale = :de

    # JavaScript files you want as :defaults (application.js is always included).
    # config.action_view.javascript_expansions[:defaults] = %w(jquery rails)

    # Configure the default encoding used in templates for Ruby 1.9.
    config.encoding = "utf-8"

    # Configure sensitive parameters which will be filtered from the log file.
    config.filter_parameters += [:password]
  end
end


lib/rqrcode_svg.rb
module RQRCode
  module SVG
    require "action_controller"
    Mime::Type.register "image/svg+xml", :svg
    
    require 'rqrcode'
    ActionController::Renderers.add :svg do |string, options|      
      qrcode = RQRCode::QRCode.new(string)
    
    
      offset = 20
      
      dimension = (qrcode.module_count*11) + (2*offset)
      
      head = <<-XML
<?xml version="1.0" standalone="yes"?>
<svg version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns:ev="http://www.w3.org/2001/xml-events" width="#{dimension}" height="#{dimension}">
      XML
      
      res = []
      qrcode.modules.each_index do |c|
        tmp = []
        qrcode.modules.each_index do |r|
          y = c*11 + offset
          x = r*11 + offset
          row = %{<rect width="11" height="11" x="#{x}" y="#{y}" style="fill:#000"/>}
          col = %{<rect width="11" height="11" x="#{x}" y="#{y}" style="fill:#fff"/>}
          if qrcode.is_dark(c,r)
            tmp << row 
          else
            tmp << col 
          end
        end 
        res << tmp.join
      end
    
      svg = (head + res.join("") + "</svg>")
      self.response_body = render_to_string(:text => svg, :template => nil)
    end
  end
end

lib/rqrcode-rails3.rb
require 'action_controller'
require 'rqrcode'
require 'rqrcode-rails3/size_calculator.rb'
require 'rqrcode-rails3/renderers/svg.rb'

module RQRCode
  Mime::Type.register "image/svg+xml", :svg  unless Mime::Type.lookup_by_extension(:svg)
  Mime::Type.register "image/png",     :png  unless Mime::Type.lookup_by_extension(:png)
	Mime::Type.register "image/jpeg",    :jpeg unless Mime::Type.lookup_by_extension(:jpeg)
  Mime::Type.register "image/gif",     :gif  unless Mime::Type.lookup_by_extension(:gif)

  extend SizeCalculator

  ActionController::Renderers.add :qrcode do |string, options|
    format = self.request.format.symbol
    size   = options[:size]  || RQRCode.minimum_qr_size_from_string(string)
    level  = options[:level] || :h

    qrcode = RQRCode::QRCode.new(string, :size => size, :level => level)
    svg    = RQRCode::Renderers::SVG::render(qrcode, options)

    data = \
    if format && format == :svg
      svg
    else
      image = MiniMagick::Image.read(svg) { |i| i.format "svg" }
      image.format format
      image.to_blob
    end

    self.response_body = render_to_string(:text => data, :template => nil)
  end
end


lib/rqrcode_rails3/svg_renderer.rb
module RQRCode
  module Renderers
    class SVG
      class << self
        def render(qrcode, options={})
          # Padding around the qrcode 
          offset = options[:offset] || 20

          # height and width dependent on offset and QR complexity
          dimension = (qrcode.module_count*11) + (2*offset)

          xml_tag   = %{<?xml version="1.0" standalone="yes"?>}
          open_tag  = %{<svg version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns:ev="http://www.w3.org/2001/xml-events" width="#{dimension}" height="#{dimension}">}
          close_tag = "</svg>"

          result = []
          qrcode.modules.each_index do |c|
            tmp = []
            qrcode.modules.each_index do |r|
              y = c*11 + offset
              x = r*11 + offset
              row = %{<rect width="11" height="11" x="#{x}" y="#{y}" style="fill:#000"/>}
              col = %{<rect width="11" height="11" x="#{x}" y="#{y}" style="fill:#fff"/>}
              if qrcode.is_dark(c,r)
                tmp << row 
              else
                tmp << col 
              end
            end 
            result << tmp.join
          end
          svg = [xml_tag, open_tag, result, close_tag].flatten.join("\n")
        end
      end
    end
  end
end


lib/rqrcode-rails3/size_calculator.rb
module RQRCode
  module SizeCalculator
    # size - seems to follow this logic
    #     # | input | modules
    #       | size  | created
    #-------|-------|--------
    #     1 |     7 |      21
    #     2 |    14 |      25 (+4)
    #     3 |    24 |      29   -
    #     4 |    34 |      33   -
    #     5 |    44 |      37   -
    #     6 |    58 |      41   -
    #     7 |    64 |      45   -
    #     8 |    84 |      49   -
    #     9 |    98 |      53   -
    #    10 |   119 |      57   -
    #    11 |   137 |      61   -
    #    12 |   155 |      65   -
    #    13 |   177 |      69   -
    #    14 |   194 |      73   -

    QR_CHAR_SIZE_VS_SIZE = [7, 14, 24, 34, 44, 58, 64, 84, 98, 119, 137, 155, 177, 194]

    def minimum_qr_size_from_string(string)
      QR_CHAR_SIZE_VS_SIZE.each_with_index do |size, index|
        return (index + 1) if string.size < size
      end

      # If it's particularly big, we'll try and create codes until it accepts
      i = QR_CHAR_SIZE_VS_SIZE.size
      begin
        i += 1
        RQRCode::QRCode.new(string, :size => i)
        return i
      rescue RQRCode::QRCodeRunTimeError
        retry
      end
    end
  end
end

lib/rqrcode-rails3/renderers/svg.rb
module RQRCode
  module Renderers
    class SVG
      class << self
        # Render the SVG from the qrcode string provided from the RQRCode gem
        #   Options:
        #   offset - Padding around the QR Code (e.g. 10)
        #   unit   - How many pixels per module (Default: 11)
        #   fill   - Background color (e.g "ffffff" or :white)
        #   color  - Foreground color for the code (e.g. "000000" or :black)

        def render(qrcode, options={})
          offset  = options[:offset].to_i || 0
          color   = options[:color]       || "000"
          unit    = options[:unit]        || 11

          # height and width dependent on offset and QR complexity
          dimension = (qrcode.module_count*unit) + (2*offset)

          xml_tag   = %{<?xml version="1.0" standalone="yes"?>}
          open_tag  = %{<svg version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns:ev="http://www.w3.org/2001/xml-events" width="#{dimension}" height="#{dimension}">}
          close_tag = "</svg>"

          result = []
          qrcode.modules.each_index do |c|
            tmp = []
            qrcode.modules.each_index do |r|
              y = c*unit + offset
              x = r*unit + offset

              next unless qrcode.is_dark(c, r)
              tmp << %{<rect width="#{unit}" height="#{unit}" x="#{x}" y="#{y}" style="fill:##{color}"/>}
            end 
            result << tmp.join
          end
          
          if options[:fill]
            result.unshift %{<rect width="#{dimension}" height="#{dimension}" x="0" y="0" style="fill:##{options[:fill]}"/>}
          end
          
          svg = [xml_tag, open_tag, result, close_tag].flatten.join("\n")
        end
      end
    end
  end
end
